nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,502 @@
+				Required:    true,
+			},
+			"bundle_upload_timeout": {
+				Type:        schema.TypeInt,
+				Description: "Timeout for uploading bundle in seconds",
+				Optional:    true,
+				Default:     bundleUploadTimeout,
+			},
+			"uc_upgrade_timeout": {
+				Type:        schema.TypeInt,
+				Description: "Timeout for upgrading upgrade coordinator in seconds",
+				Optional:    true,
+				Default:     ucUpgradeTimeout,
+			},
+			"precheck_timeout": {
+				Type:        schema.TypeInt,
+				Description: "Timeout for executing pre-upgrade checks in seconds",
+				Optional:    true,
+				Default:     precheckTimeout,
+			},
+			"failed_prechecks": {
+				Type:        schema.TypeList,
+				Description: "List of failed prechecks for the upgrade, including both warnings and errors",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:        schema.TypeString,
+							Description: "Type of the precheck failure (warning or error)",
+							Computed:    true,
+						},
+						"id": {
+							Type:        schema.TypeString,
+							Description: "ID of failed precheck",
+							Computed:    true,
+						},
+						"message": {
+							Type:        schema.TypeString,
+							Description: "Message of the failed precheck",
+							Computed:    true,
+						},
+						"needs_ack": {
+							Type:        schema.TypeBool,
+							Description: "Boolean value which identifies if acknowledgement is required for the precheck",
+							Computed:    true,
+						},
+						"needs_resolve": {
+							Type:        schema.TypeBool,
+							Description: "Boolean value identifies if resolution is required for the precheck",
+							Computed:    true,
+						},
+					},
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceNsxtUpgradePrepareCreate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	err := prepareForUpgrade(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	d.SetId(id)
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func prepareForUpgrade(d *schema.ResourceData, m interface{}) error {
+	// 1. Upload upgrade bundle and wait for upload to complete
+	err := uploadPrecheckAndUpgradeBundle(d, m)
+	if err != nil {
+		return logAPIError("Failed to upload bundle", err)
+	}
+	// 2. Accept eula
+	err = acceptUserAgreement(d, m)
+	if err != nil {
+		return err
+	}
+	// 3. Upgrade UC and check for its upgrade status
+	err = upgradeUc(d, m)
+	if err != nil {
+		return logAPIError("Failed to upgrade Upgrade Coordinator", err)
+	}
+	return nil
+}
+
+func precheckNeeded(m interface{}) (bool, error) {
+	connector := getPolicyConnector(m)
+	summaryClient := upgrade.NewSummaryClient(connector)
+	summary, err := summaryClient.Get()
+	if err != nil {
+		return false, err
+	}
+	var bundleURL string
+	var ucUpgraded bool
+	var upgradeStatus string
+	if summary.UpgradeBundleFileName != nil {
```

Comment:
If `UpgradeBundleFileName` is nil, seems like we need to skip precheck as well?

Reply:
Adjusted the if condition to `if summary.UpgradeBundleFileName == nil || len(*summary.UpgradeBundleFileName) == 0`
same for the other 2 conditions